src/connector.c

```diff
@@ -0,0 +1,319 @@
+// you may not use this file except in compliance with the License.
+//
+// You may obtain a copy of the License at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <tlsuv/connector.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include "util.h"
+#include "um_debug.h"
+
+
+#if _WIN32
+#include "win32_compat.h"
+#include <winsock2.h>
+#define ioctl ioctlsocket
+#define get_error() WSAGetLastError()
+#define write(s,b,z) send(s, b, z, 0)
+#define read(s,b,z) recv(s,b,z,0)
+#define SHUT_RDWR SD_BOTH
+#else
+#define get_error() errno
+#define closesocket(s) close(s)
+#include <poll.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <string.h>
+
+#endif
+
+
+struct conn_req_s {
+    union {
+        uv_getaddrinfo_t resolve;
+        uv_work_t connect;
+    };
+    void *ctx;
+    tlsuv_connect_cb cb;
+
+    uv_os_sock_t sock;
+    int error;
+    int cancel;
+};
+
+static tlsuv_connector_req default_connect(uv_loop_t *l, const tlsuv_connector_t *self,
+                                           const char *host, const char *port, tlsuv_connect_cb cb, void *ctx);
+static void default_cancel(tlsuv_connector_req req);
+static tlsuv_connector_req proxy_connect(uv_loop_t *l, const tlsuv_connector_t *self,
+                                         const char *host, const char *port, tlsuv_connect_cb cb, void *ctx);
+static void proxy_cancel(tlsuv_connector_req req);
+
+// prevent freeing default connector
+static void default_connector_free(void *self){
+    (void)self;
+}
+
+static tlsuv_connector_t default_connector = {
+        .connect = default_connect,
+        .cancel = default_cancel,
+        .free = default_connector_free,
+};
+static const tlsuv_connector_t *global_connector = &default_connector;
+
+struct tlsuv_proxy_connector_s {
+    tlsuv_connect connect;
+    void (*cancel)(tlsuv_connector_req);
+    void (*free)(tlsuv_connector_t *self);
+    
+    tlsuv_proxy_t type;
+    char *host;
+    char *port;
+};
+
+void tlsuv_set_global_connector(const tlsuv_connector_t *c) {
+    global_connector = c ? c : &default_connector;
+}
+
+const tlsuv_connector_t* tlsuv_global_connector() {
+    return global_connector;
+}
+
+static void on_close(uv_handle_t *h) {
+    struct conn_req_s *req = h->data;
+    free(req);
+}
+
+static void def_connect_work(uv_work_t *w) {
+    struct conn_req_s *r = container_of(w, struct conn_req_s, connect);
+
+    struct pollfd fds[] = {
+            {
+                    .fd = r->sock,
+                    .events = POLLOUT,
+                    .revents = 0,
+            }
+    };
+#if _WIN32
+    WSAPoll(fds, 1, -1);
+#else
+    poll(fds, 1, -1);
+#endif
+
```

Comment:
Could these polls actually block indefinitely (timeout=-1)?

And should the return status from poll be checked before using getsockopt?